lib/org/chromium/webidl/IDLFragmentExtractor.js

```diff
@@ -21,96 +22,179 @@ foam.CLASS({
+      factory: function() {
+        return this.TagStack.create({ isExcluded: this.isExcluded });
+      },
+    },
+    {
+      name: 'tagMatching_',
+      documentation: 'Used to determine if tag matching is currently performed.',
+      value: true,
     },
   ],
 
   methods: [
-    function extract() {
-      var lexer = this.HTMLLexer.create();
-      var OPEN = this.TagType.OPEN.name;
-      var CLOSE = this.TagType.CLOSE.name;
-      var extractAttr = function(node, attrName) {
+    {
+      name: 'extractAttr',
+      documentation: 'Extracts given attribute from tag.',
+      code: function(tag, attrName) {
         var retVal = [];
-        node.attributes.forEach(function(attr) {
+        tag.attributes.forEach(function(attr) {
           if (attr.name === attrName) {
             retVal = retVal.concat(attr.value.split(' '));
           }
         });
         return retVal;
-      };
-
-      var results = lexer.parseString(this.file.contents).value;
-      if (!results) throw new Error("IDL Parse was not successful.");
-
-      var idlFragments = [];
-      var tagStack = [];            // Used for tag matching.
-      var excludeStack = [];        // Used for tracking excluded (example/note) tags.
-      var tagMatching = true;       // Set whe not inside a pre tag of interest.
-      var content = '';             // Used to group together related content.
-      for (var i = 0; i < results.length; i++) {
-        var item = results[i];
-        var isTag = lexer.Tag.isInstance(item);
-
-        // FUTURE: Handle nested pre tags.
-        // As of this writing, there has not been any IDL fragments
-        // that has been found within nested pre tags.
-        if (!tagMatching) {
-          // Ignoring all tags. Only extracting text content within pre tags.
-          if (isTag) {
-            if (item.nodeName === 'pre') {
-              tagMatching = true;
-              tagStack.pop();
-              idlFragments.push(lexer.lib.unescapeString(content));
-              content = '';
+      },
+    },
+    {
+      name: 'isExcluded',
+      documentation: 'Determines if tag has excluded classes.',
+      code: function(tag) {
+        var cls = [];
+        tag.attributes.forEach(function(attr) {
+          if (attr.name === 'class') {
+            cls = cls.concat(attr.value.split(' '));
+          }
+        });
+        return cls.includes('example') || cls.includes('note');
+      },
+    },
+    {
+      name: 'openTagAction',
+      documentation: 'Action performed when an open HTML tag is encountered.',
+      code: function(tag) {
+        var top = this.tagStack_.top;
+        var excludedTop = this.tagStack_.excludedTop;
+
+        var classes = this.extractAttr(tag, 'class');
+        var isIDL = classes && classes.includes('idl');
+
+        // HACK: Top level tags can be unmatched as well. One noticeable example
+        // is <p>. However, <div> cannot nest within <p>, so if we see an open
```

Comment:
As per the [HTML 5 specification](http://w3c.github.io/html/single-page.html#the-p-element), only the elements listed can be nested within a `<p>` tag. `<div>` is not one of them and a couple of the specs rely on this property to close off the `<p>` tag properly